OpenRA.Mods.Common/Activities/Air/Fly.cs

```diff
@@ -37,36 +37,37 @@ public Fly(Actor self, Target t, WDist minRange, WDist maxRange)
 			this.minRange = minRange;
 		}
 
-		public static void FlyToward(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude)
+		public static bool FlyToward(Actor self, Aircraft aircraft, int desiredFacing, WDist desiredAltitude,
+			bool moveVerticalOnly = false, bool flyBackward = false, bool isFlyCircle = false, int turnSpeedOverride = -1)
 		{
 			desiredAltitude = new WDist(aircraft.CenterPosition.Z) + desiredAltitude - self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
 
-			var move = aircraft.FlyStep(aircraft.Facing);
 			var altitude = aircraft.CenterPosition.Z;
+			if (moveVerticalOnly && altitude == desiredAltitude.Length)
+				return false;
+
+			var move = WVec.Zero;
+			if (!moveVerticalOnly)
+			{
+				// If FlyToward is called from FlyCircle, the aircraft needs to use aircraft.Facing even if it CanHover,
+				// otherwise it will circle sideways.
+				if (aircraft.Info.CanHover && !isFlyCircle)
+					move = aircraft.FlyStep(flyBackward ? -desiredFacing : desiredFacing);
```

Comment:
IMO it would be cleaner to define a `var direction = flyBackwards ? -1 : 1` and then use `FlyStep(direction * facing)` here and below.